### ISSUE TITLE
Replace psl dependency with tldts

### ISSUE THREAD

**Issue by isaacl** [salesforce/tough-cookie#338]
The [psl](https://github.com/lupomontero/psl) project seems to be [dead](https://github.com/lupomontero/psl/pulls).

It's generating a [punycode deprecation warning](#324) despite an 18mo old open pull request fixing this issue.  tough-cookie incurs a serious [performance bottleneck](#239) due to its use of psl.

tldts is [active](https://github.com/remusao/tldts/pulls?q=is%3Apr+is%3Amerged+), and [much, much more performant](https://remusao.github.io/posts/tldts-benchmarks.html). 

Can you fix this, please?

**Comment by colincasey** [comment]
Thanks for opening this issue @isaacl. We have discussed a replacement for `psl` and `tldts` look decent.

**Comment by obecker** [comment]
@colincasey I think you should consider also removing `"@types/psl": "^1.1.3"` from the `devDependencies`.

**Comment by colincasey** [comment]
Thanks for calling that out @obecker.  It will be removed with #406.

**Comment by skagedal** [comment]
Thank you for this fix! It seems it didn't make it into the 4.1.4 release. Will there be a new release soon?

**Comment by colincasey** [comment]
@skagedal this won't be backported to 4.1.4. It's only available in our v5 release candidate for now ([5.0.0-rc.1](https://www.npmjs.com/package/tough-cookie/v/5.0.0-rc.1)).

The upcoming v5 release is primarily an internal uplift that includes TypeScript support and improved documentation so it should be a seamless upgrade but feel free to try out the release candidate ahead of time and report any issues 😄

**Comment by skagedal** [comment]
Ah, thanks, I had missed that release candidate! I tested it out in our project and indeed, PSL is now gone along with those particular `punycode` warnings. Good work.

**Comment by iambumblehead** [comment]
the punycode warning for the last year or so is very annoying

**Comment by JessicaSachs** [comment]
Hey, I wrote a bit of a book here for others to find. Please feel free to reference it as people complain about trying to suppress the warning from `punycode`.

> @skagedal this won't be backported to 4.1.4. It's only available in our v5 release candidate for now ([5.0.0-rc.1](https://www.npmjs.com/package/tough-cookie/v/5.0.0-rc.1)).
> 
> The upcoming v5 release is primarily an internal uplift that includes TypeScript support and improved documentation so it should be a seamless upgrade but feel free to try out the release candidate ahead of time and report any issues 😄

The `5.0.0-rc.1` release doesn't seem to remove the dependency on punycode **BUT THE 5.0.0-rc.4 version does**.

## Steps to override transitive `tough-cookie` deps to use `5.0.0-rc.4`

If you have a dependency on `tough-cookie` from a test-only package like `jsdom` or `msw`, here's how you override the dependency to fall through to `tough-cookie`'s pre-release version. I suggest doing a fine-grained control to scope overrides to the build-only or test-only packages where your changes are more-or-less guaranteed to be harmless.

### 1. Find the deps (`npm ls punycode`)
Detect where the transitive dep is from using `npm ls punycode`. If you're using `msw` (v2.3.5 at time of writing), it depends on `tough-cookie`. You'll see something like this output when running `npm ls punycode`. 

<img width="400" alt="Screenshot 2024-09-05 at 6 19 05 PM" src="https://github.com/user-attachments/assets/d53b8caa-a2d7-4c5e-aa7d-3bc631a3ce1e">

### 2. Trash your lockfiles and node_modules

### 3. Setup overrides in `package.json`
Start forcing resolutions using npm overrides. I suggest doing this in a way that you're sure is safe for your codebase. You **can** just say `"tough-cookie": "5.0.0-rc.4"` but I wouldn't risk using an rc version in production without significant testing.

**Example fixed package.json**:
```json5
{
  "dependencies": {
    "msw": "latest"
  },
  "overrides": {
    "msw": {
      "tough-cookie": "5.0.0-rc.4"
    }
  }
}
```

#### 4. `npm install`
If you do it right:
* `npm ls punycode` should come up empty

<img width="400" alt="Screenshot 2024-09-05 at 6 40 17 PM" src="https://github.com/user-attachments/assets/9c9aa868-7df7-4b35-bc6c-8c7e0426f0da">

* looking in your `node_modules` for `tough-cookie` should show the proper version number.
<img width="400" alt="Screenshot 2024-09-05 at 6 40 01 PM" src="https://github.com/user-attachments/assets/c99246b6-ff9d-413f-90c2-a69101d5863a">

**Other package managers:**
* [pnpm overrides](https://pnpm.io/package_json#pnpmoverrides)
* [bun overrides](https://bun.sh/docs/install/overrides)
* [deno overrides (not implemented?)](https://github.com/denoland/deno/issues/20614)
* [yarn resolutions (good luck)](https://classic.yarnpkg.com/lang/en/docs/selective-version-resolutions/)

**Comment by MikeMcC399** [comment]
@JessicaSachs 

I'm seeing [tough-cookie@5.0.0](https://github.com/salesforce/tough-cookie/releases/tag/v5.0.0) now released.
